File: day9/haskell/part1.hs
import System.Environment (getArgs)

splitOn :: Char -> String -> [String]
splitOn _ [] = [""]
splitOn delim (c:cs)
  | c == delim = "" : rest
  | otherwise  = (c : head rest) : tail rest
  where rest = splitOn delim cs

parsePoint :: String -> (Int, Int)
parsePoint s = (read x, read y)
  where [x, y] = splitOn ',' s

rectangleArea :: (Int, Int) -> (Int, Int) -> Int
rectangleArea (x1, y1) (x2, y2) = width * height
  where
    width = abs (x2 - x1) + 1
    height = abs (y2 - y1) + 1

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "input.txt" else head args
  content <- readFile filename
  let tiles = map parsePoint $ lines content

      -- All pairs of tiles
      pairs = [(tiles !! i, tiles !! j)
              | i <- [0..length tiles - 1]
              , j <- [i+1..length tiles - 1]]

      -- Calculate area for each pair
      areas = map (uncurry rectangleArea) pairs

      maxArea = maximum areas

  print maxArea

File: day9/go/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Point struct {
	x, y int
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}
	file, _ := os.Open(filename)
	defer file.Close()

	var tiles []Point
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		tiles = append(tiles, Point{x, y})
	}

	maxArea := 0

	// Try all pairs of tiles as opposite corners
	for i := 0; i < len(tiles); i++ {
		for j := i + 1; j < len(tiles); j++ {
			width := abs(tiles[j].x-tiles[i].x) + 1
			height := abs(tiles[j].y-tiles[i].y) + 1
			area := width * height

			if area > maxArea {
				maxArea = area
			}
		}
	}

	fmt.Println(maxArea)
}
